errors/errors.go

```diff
@@ -15,209 +15,310 @@ package errors
-	}
-
-	// Now apply any requested options
-	// to override
-	for _, option := range options {
-		option(de)
+	kitError := &Error{
+		Details:  make([]proto.Message, 0),
+		GrpcCode: grpcCode,
+		HttpCode: httpCode,
+		Message:  message,
+		Tag:      tag,
 	}
 
-	return de
+	return kitError
 }
 
 // Error implements the error interface.
 func (e *Error) Error() string {
-	if e != nil && e.err != nil {
-		return e.err.Error()
+	if e != nil {
+		return e.String()
 	}
 	return ""
 }
 
-// Unwrap implements the error unwrapping interface.
-func (e *Error) Unwrap() error {
-	if e == nil {
-		return nil
-	}
-	return e.err
+// String returns the string representation, useful for debugging.
+func (e Error) String() string {
+	return fmt.Sprintf(errStringFormat, e.GrpcCode.String(), e.Message)
 }
 
-// Description returns the description of the error.
-func (e *Error) Description() string {
-	if e == nil {
-		return ""
-	}
-	if e.description != "" {
-		return e.description
+// WithResourceInfo used to pass ResourceInfo to the Error struct.
+func (e *Error) WithResourceInfo(resourceType string, resourceName string, owner string, description string) *Error {
+	resourceInfo := &errdetails.ResourceInfo{
+		ResourceType: resourceType,
+		ResourceName: resourceName,
+		Owner:        owner,
+		Description:  description,
 	}
-	return e.err.Error()
-}
 
-// WithErrorReason used to pass reason and
-// grpcStatus code to the Error struct.
-func WithErrorReason(reason string, grpcStatusCode codes.Code) Option {
-	return func(err *Error) {
-		err.reason = reason
-		err.grpcStatusCode = grpcStatusCode
-		err.httpCode = grpccodes.HTTPStatusFromCode(grpcStatusCode)
-	}
-}
+	e.Details = append(e.Details, resourceInfo)
 
-// WithResourceInfo used to pass ResourceInfo to the Error struct.
-func WithResourceInfo(resourceInfo *ResourceInfo) Option {
-	return func(e *Error) {
-		e.resourceInfo = resourceInfo
-	}
+	return e
 }
 
-// WithDescription used to pass a description
-// to the Error struct.
-func WithDescription(description string) Option {
-	return func(e *Error) {
-		e.description = description
+// WithErrorInfo adds error information to the Error struct.
+func (e *Error) WithErrorInfo(reason string, metadata map[string]string) *Error {
+	errorInfo := &errdetails.ErrorInfo{
+		Domain:   ErrMsgDomain,
+		Reason:   reason,
+		Metadata: metadata,
 	}
-}
+	e.Details = append(e.Details, errorInfo)
 
-// WithMetadata used to pass a Metadata[string]string
-// to the Error struct.
-func WithMetadata(md map[string]string) Option {
-	return func(e *Error) {
-		e.metadata = md
-	}
+	return e
 }
 
-func newErrorInfo(reason string, md map[string]string) *errdetails.ErrorInfo {
-	return &errdetails.ErrorInfo{
-		Domain:   errorInfoDefaultDomain,
-		Reason:   reason,
-		Metadata: md,
-	}
+// WithVars returns a copy of the error with the message going through fmt.Sprintf with the arguments passed to this method.
+func (e *Error) WithVars(a ...any) *Error {
+	eCopy := *e
```

Comment:
Not 100% but I don't think this copies the entire object. At least the slice that is contained (`Details`) may not be copied

Reply:
this func will be removed entirely